IBC.md

```diff
@@ -0,0 +1,412 @@
+[cw20 tokens](https://github.com/CosmWasm/cosmwasm-plus/tree/v0.6.0-beta1/packages/cw20)
+and sends those to a remote chain, as well as receiving the tokens back and
+releasing the original cw20 token to a new owner. It does not (yet) allow
+minting coins originating from the remote chain. I recommend opening up the
+source code for that contract and refering to it when you want a concrete
+example for anything discussed below.
+
+In order to enable IBC communication, a contract must expose the following 6
+entry points. Upon detecting such an "IBC-Enabled" contract, the
+[wasmd runtime](https://github.com/CosmWasm/wasmd) will automatically bind a
+port for this contract (`wasm.<contract-address>`), which allows a relayer to
+create channels between this contract and another chain. Once channels are
+created, the contract will process all packets and receipts.
+
+### Channel Lifecycle
+
+You should first familiarize yourself with the
+[4 step channel handshake protocol](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#channel-lifecycle-management)
+from the IBC spec. After realizing that it was 2 slight variants of 2 steps, we
+simplified the interface for the contracts. Each side will receive 2 calls to
+establish a new channel, and returning an error in any of the steps will abort
+the handshake. Below we will refer to the chains as A and B - A is where the
+handshake initialized at.
+
+#### Channel Open
+
+The first step of a handshake on either chain is `ibc_channel_open`, which
+combines `ChanOpenInit` and `ChanOpenTry` from the spec. The only valid action
+of the contract is to accept the channel or reject it. This is generally based
+on the ordering and version in the `IbcChannel` information, but you could
+enforce other constraints as well:
+
+```rust
+#[entry_point]
+/// enforces ordering and versioning constraints
+pub fn ibc_channel_open(deps: DepsMut, env: Env, channel: IbcChannel) -> StdResult<()> { }
+```
+
+This is the
+[IbcChannel structure](https://github.com/CosmWasm/cosmwasm/blob/v0.14.0-beta4/packages/std/src/ibc.rs#L70-L81)
+used heavily in the handshake process:
+
+```rust
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct IbcChannel {
+    pub endpoint: IbcEndpoint,
+    pub counterparty_endpoint: IbcEndpoint,
+    pub order: IbcOrder,
+    pub version: String,
+    /// CounterpartyVersion can be None when not known this context, yet
+    pub counterparty_version: Option<String>,
+    /// The connection upon which this channel was created. If this is a multi-hop
+    /// channel, we only expose the first hop.
+    pub connection_id: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct IbcEndpoint {
+    pub port_id: String,
+    pub channel_id: String,
+}
+```
+
+Note that neither `counterparty_version` nor `counterparty_endpoint` is set in
+`ibc_channel_open` for chain A. Chain B should enforce any
```

Comment:
How is this possible with non-optional type `counterparty_endpoint: IbcEndpoint`?

Reply:
I think it just gets "" fields. You are right, we should make this optional if it is not present, but maybe that would confused all calls but one where it is optional (I can use `IbcChannelWithOptionalCounterparty` maybe for this?)

Reply:
We pass in the counterparty endpoint on init as well: https://github.com/CosmWasm/wasmd/blob/master/x/wasm/ibc.go#L48

I am not sure if these are set in the sdk before calling us

Reply:
After digging more in the sdk code, I found that counterparty.port is set here (in `ChannelOpenInit`), but counterparty.channel is not set.

https://github.com/cosmos/cosmos-sdk/blob/88e03e4f4056e8614847fd6adf17ce9252b6a89b/x/ibc/core/04-channel/types/msgs.go#L41-L60

In `ChannelOpenTry`, the counterparty channel must be set (ignore that comment, the code shows otherwise): https://github.com/cosmos/cosmos-sdk/blob/88e03e4f4056e8614847fd6adf17ce9252b6a89b/x/ibc/core/04-channel/types/msgs.go#L131-L134

Since we collapse these 2 into one call (to limit the number of exports and duplicated code), I am not sure the best way to handle this.

Making `channel` optional in all `IbcEndpoint` would be overkill and complicate other logic. Using `counterparty: IbcEndpointWithOptionalChannel` for a custom type for `ibc_channel_open` could work, along with a comment that `counterparty_version` and `counterparty.channel` are both set or both unset.

Any other ideas?

Reply:
Do we need `counterparty_endpoint.port_id` when `counterparty_version` and `counterparty_version.channel` is unset? If not, I think

```
pub counterparty_endpoint: Option<IbcEndpoint>,
```

would be solid.

Reply:
Some protocols (like ics27 currently) try to enforce the remote port.
Few if any care about the remote channel_id

Reply:
Hmm, okay. Then I don't have a great idea. In this case it starts feeling wrong to use the same type for two different calls because we don't have dynamic typing available here.